Migrate Ansible content after upgrade to RHEL8 (#1491)
theforeman/foreman-documentation: guides/common/modules/ref_migrating-ansible-content.adoc

[id="migrating-ansible-content_{context}"]
= Migrating Ansible Content

> The upgrade from {RHEL} 7 to {RHEL} 8 includes an upgrade from Ansible Engine 2.9 to Ansible Core 2.12.

Author: @maximiliankolb Do you think it would be useful to generalize this procedure for EL systems?

Author: @ekohl Thanks!

If you have custom Ansible content such as playbooks, job templates inside REX, roles and collections on disk, and you relied on modules being delivered by the Ansible RPM on {Project}, you have to take additional steps to adapt your Ansible installation or migrate your Ansible content.

> Ansible Core reduces the amount of non-essential modules shipped in the core installation.

Reviewer: This can be stronger. Ansible Core has removed all non-essential modules.

In terms of FQCN notation `namespace.collection.module`, you can continue to use `ansible.builtin.*`, but everything else is missing in Ansible Core.
That means you will be no longer able to use non-builtin Ansible modules as you were used to and you have to get them from somewhere else, eventually.

> Current options to resolve this situation:

Author: @ekohl Btw, what solutions can I use on non-RH EL?

[id="migrating-ansible-content_{context}"]
= Migrating Ansible Content

The upgrade from {RHEL} 7 to {RHEL} 8 includes an upgrade from Ansible Engine 2.9 to Ansible Core 2.12.

If you have custom Ansible content such as playbooks, job templates inside REX, roles and collections on disk, and you relied on modules being delivered by the Ansible RPM on {Project}, you have to consider migrating your Ansible content.

> If you are using FQCN notation `namespace.collection.module` in your content, you can continue to use `ansible.builtin.*`, but everything else is missing in Ansible Core.

Author: This sounds like it depends on the notation, which I think is not the case, may need further rewrite

That means you will be no longer able to use non-builtin Ansible modules as you were used to and you have to get them from somewhere else, eventually.

Current options to resolve this situation:

* You download Ansible Galaxy from the AppStream RHEL 8 repository.

> If you need support regarding `/usr/bin/ansible-playbook`, you can reach out to Red Hat support, but we do not provide support for the content you try to execute with it.

Author: @ekohl Do you know the answer?

> If you need support regarding `/usr/bin/ansible-playbook`, you can reach out to Red Hat support, but we do not provide support for the content you try to execute with it.

Reviewer 2: While this statement is technically true, it feels miss-placed here?
I would probably either avoid it all-together, or mention something like this in the "you rewrite" step below.

* You buy Ansible subscription, configure `ansible-galaxy` to talk to `cloud.redhat.com` and use it to download supported collections from there.
If you need support, you reach out to the Ansible team.
* You rewrite your templates, roles, etc.

To find out what content is affected, you can run:
[options="nowrap" subs="+quotes,attributes"]
----
# ansible-lint -t only-builtins -p _path/to/your/playbook/play.yml_ ...
----

> You need `ansible-lint` of version X.Y or newer. // TBD

Author: Does the user have to worry about the version after the upgrade?

[id="migrating-ansible-content_{context}"]
= Migrating Ansible Content

The upgrade from {EL} 7 to {EL} 8 includes an upgrade from Ansible Engine 2.9 to Ansible Core 2.12.

If you have custom Ansible content such as playbooks, job templates inside REX, roles and collections on disk, and you relied on modules being delivered by the Ansible RPM on {Project}, you have to take additional steps to adapt your Ansible installation or migrate your Ansible content.

Ansible Core removed all non-essential modules shipped in the core installation.
In terms of FQCN notation `namespace.collection.module`, you can continue to use `ansible.builtin.*`, but everything else is missing in Ansible Core.
That means you will be no longer able to use non-builtin Ansible modules as you were used to and you have to get them from somewhere else, eventually.

Current options to resolve this situation:

> * You download Ansible Galaxy from the AppStream RHEL 8 repository.

Reviewer 2: The `ansible-core` package (which contains the `ansible-galaxy` binary) is automatically installed when you upgrade from EL7 to EL8 or when you deploy fresh on EL8, so this step is unnecessary.

Author: Any EL or RHEL specifically?

Author: Seriously, if they get galaxy, which contains the additional modules I presume, doesn't it render this whole module unnecessary? I'm confused.

Reviewer 2: ansible-galaxy is the tool, Ansible Galaxy (aka galaxy.ansible.com) is the online service the tool talks to

Author: Okay, but I still don't understand what it implies for the text that is here and how I'm supposed to improve it. I'll talk to you on IRC.

If you need support regarding `/usr/bin/ansible-playbook`, you can reach out to Red Hat support, but we do not provide support for the content you try to execute with it.
* You buy an Ansible subscription, configure `ansible-galaxy` to talk to `cloud.redhat.com` and use it to download supported collections from there.

> If you need support, you reach out to the Ansible team.

Reviewer 2: Upstream users can get most of the same collections from `galaxy.ansible.com`, minus the enterprise support.
So can Satellite customers, they just need to be aware that that content is then not supported by anybody officially (the support-ability of the rest of the setup remains, just the content is unsupported).

* You rewrite your templates, roles, etc.

> To find out what content is affected after the upgrade, you can run:
> [options="nowrap" subs="+quotes,attributes"]
> ----
> # ansible-lint -t only-builtins -p _path/to/your/playbook/play.yml_ ...
> ----

Reviewer 2: this rule (only-builtins) was added in ansible-lint 6.1.0 (and then there was an annoying false-postive bug, which only got fixed in 6.3.0), which is not packaged at all yet for any platform we support
so while yes, this is a good way to check it, we cannot recommend it to users, as they have no way of obtaining a version of the tool that would make it possible.

additionally, one would need to call it with `--enable-list`, not with `-t`, as it's an optional rule.

```suggestion
To find out what content is affected after the upgrade, you can run:
[options="nowrap" subs="+quotes,attributes"]
----
# ansible-lint --enable-list only-builtins -p _path/to/your/playbook/play.yml_ ...
----
```

[id="migrating-ansible-content_{context}"]
= Migrating Ansible Content

The upgrade from {EL} 7 to {EL} 8 includes an upgrade from Ansible Engine 2.9 to Ansible Core 2.12.

If you have custom Ansible content such as playbooks, job templates inside REX, roles and collections on disk, and you relied on modules being delivered by the Ansible RPM on {Project}, you have to take additional steps to adapt your Ansible installation or migrate your Ansible content.

Ansible Core removed all non-essential modules shipped in the core installation.
In terms of FQCN notation `namespace.collection.module`, you can continue to use `ansible.builtin.*`, but everything else is missing in Ansible Core.
That means you will be no longer able to use non-builtin Ansible modules as you were used to and you have to get them from somewhere else, eventually.

Current options to resolve this situation:

* You can download and install additional community-maintained modules from Galaxy server.
For more information, see https://galaxy.ansible.com/docs/using/installing.html[Installing content] in _Ansible Galaxy Documentation_.

> ifdef::satellite[]

Reviewer 3: ```suggestion
ifdef::satellite,orcharhino[]
```

+
Note that {Team} does not provide support for this content.
endif::[]
ifdef::satellite[]
* If you have a subscription for {Team} Automation Hub, you can configure your `ansible-galaxy` to talk to Automation Hub server and download content from there.
That content is supported by {Team}.
For more information on configuring Automation Hub connection for `ansible-galaxy`, see
https://access.redhat.com/documentation/en-us/red_hat_ansible_automation_platform/2.1/html/getting_started_with_automation_hub/proc-configure-automation-hub-server
endif::[]
* You rewrite your templates, roles, etc.
ifdef::satellite[]
Note that {Team} does not provide support for content that you maintain yourself.
endif::[]

////

> // Not packaged with RHEL yet

Reviewer 3: Maybe tell users to install from pypi? https://pypi.org/project/ansible-lint/

I think this is indeed a very valuable information, so maybe add `ifndef::satellite[]` or tell users to lint on their local machine? Or maybe using a container?

To find out what content is affected after the upgrade, you can run:
[options="nowrap" subs="+quotes,attributes"]
----

> # ansible-lint --enable-list only-builtins -p _path/to/your/playbook/play.yml_ ...

Reviewer 3: ```suggestion
# ansible-lint --enable-list only-builtins -p _path/to/your/playbook/play.yml_
```

[id="migrating-ansible-content_{context}"]
= Migrating Ansible Content

The upgrade from {EL} 7 to {EL} 8 includes an upgrade from Ansible Engine 2.9 to Ansible Core 2.12.

> If you have custom Ansible content such as playbooks, job templates inside REX, roles and collections on disk, and you relied on modules being delivered by the Ansible RPM on {Project}, you have to take additional steps to adapt your Ansible installation or migrate your Ansible content.

Reviewer 3: ```suggestion
If you have custom Ansible content such as playbooks, job templates inside REX, roles, and collections on disk, and you rely on modules being delivered by the Ansible RPM on {Project}, you have to take additional steps to adapt your Ansible installation or migrate your Ansible content.
```

> Ansible Core removed all non-essential modules shipped in the core installation.

Reviewer 3: please reword.

Maybe "Ansible Core only contains essential modules"?

In terms of FQCN notation `namespace.collection.module`, you can continue to use `ansible.builtin.*`, but everything else is missing in Ansible Core.

> That means you will be no longer able to use non-builtin Ansible modules as you were used to and you have to get them from somewhere else, eventually.

Reviewer 3: Please reword. I think "somewhere else" sounds a bit off.

> Current options to resolve this situation:

Reviewer 3: ```suggestion
You have the following options to get Ansible content that is no longer part of Ansible itself:
```

* You can obtain additional community-maintained collections that provide the removed functionality from Ansible Galaxy.

> For more information, see https://docs.ansible.com/ansible/devel/user_guide/collections_using.html#installing-collections[Installing collections] in _Ansible Galaxy Documentation_.

Reviewer 3: linking to devel is probably good for Foreman "master" branch, but all other branches ideally link to a specific version if possible.

Reviewer 2: `s/devel/latest/` then, it points to "the latest stable"?

Reviewer 2: Or you can explicitly use the core docs for 2.12 at https://docs.ansible.com/ansible-core/2.12/user_guide/collections_using.html

Reviewer 3: Maybe we can use an attribute for this and the set it per Foreman version? Or at least differeniate between versions between EL7 and EL8.

Reviewer 2: EL7 is no more (this doc is explicitly about moving things when coming to EL8), and EL8 is currently 2.12-exclusive, so I don't see a big benefit _right now_.

Reviewer 3: Yes, my bad!

Then "latest" seems to be a good choice!

ifdef::satellite[]
+
Note that {Team} does not provide support for this content.

> endif::[]
> ifdef::satellite[]

Reviewer 3: ```suggestion
```

* If you have a subscription for https://www.ansible.com/products/automation-hub[Red Hat Automation Hub], you can configure your `ansible-galaxy` to talk to Automation Hub server and download content from there.
That content is supported by {Team}.

> For more information on configuring Automation Hub connection for `ansible-galaxy`, see
> https://access.redhat.com/documentation/en-us/red_hat_ansible_automation_platform/2.1/html/getting_started_with_automation_hub/proc-configure-automation-hub-server[Configuring Red Hat automation hub as the primary source for content].

Reviewer 3: ```suggestion
For more information on configuring Automation Hub connection for `ansible-galaxy`, see https://access.redhat.com/documentation/en-us/red_hat_ansible_automation_platform/2.1/html/getting_started_with_automation_hub/proc-configure-automation-hub-server[Configuring Red Hat automation hub as the primary source for content].
```

endif::[]

> * You rewrite your templates, roles, etc.

Reviewer 3: Maybe "You can rewrite your Ansible roles and affected templates."

ifdef::satellite[]
Note that {Team} does not provide support for content that you maintain yourself.
endif::[]

[NOTE]
====

> If you want to download and install Ansible content on {SmartProxy} that does not have a connection to an external Galaxy server, then you must pass the content through {ProjectServer} instead of using the URL of the Galaxy server in the configuration on the {SmartProxy} directly:

Reviewer 3: ```suggestion
If you want to download and install Ansible content on {SmartProxy} that does not have a connection to an external Ansible Galaxy server, then you must pass the content through {ProjectServer} instead of using the URL of the Ansible Galaxy server in the configuration on the {SmartProxy} directly:
```

> . Sync the content from a Galaxy server to a custom repository on your {ProjectServer}.

Reviewer 3: ```suggestion
. Sync the content from an Ansible Galaxy server to a custom repository on your {ProjectServer}.
```

> . Configure Ansible on the {SmartProxy} to download the content from the {ProjectServer}.

Reviewer 3: ```suggestion
. Configure Ansible on your {SmartProxy} to download the content from {ProjectServer}.
```

====

ifndef::satellite[]
// Not packaged with RHEL8 yet
To find out what content is affected after the upgrade, you can run:

> [options="nowrap" subs="+quotes,attributes"]

Reviewer 3: ```suggestion

[options="nowrap" subs="+quotes,attributes"]
```

Not 100% if the newline is necessary. Please check if it works without.

----
# ansible-lint --enable-list only-builtins -p _path/to/your/playbook/play.yml_
----

You need `ansible-lint` 6.1.0 or newer.

> You can install it from pypi: https://pypi.org/project/ansible-lint/.

Reviewer 3: ```suggestion
You can install it from https://pypi.org/project/ansible-lint/[pypi.org].
```

endif::[]

ifdef::satellite[]
.Additional resources
* https://www.redhat.com/en/blog/updates-using-ansible-rhel-86-and-90[Updates to using Ansible in RHEL 8.6 and 9.0]
* https://access.redhat.com/articles/6393361[Using Ansible in RHEL 8.6 and later]

> * https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8-beta/html-single/8.6_release_notes/index[Release Notes for Red Hat Enterprise Linux 8.6]

Reviewer 2: ```suggestion
* https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html-single/8.6_release_notes/index[Release Notes for Red Hat Enterprise Linux 8.6]
```

let's link to final RNs not the ones for the beta ;-)

Reviewer 2: this was the last comment, I swear!

[id="migrating-ansible-content_{context}"]
= Migrating Ansible Content

The upgrade from {EL} 7 to {EL} 8 includes an upgrade from Ansible Engine 2.9 to Ansible Core 2.12.

If you have custom Ansible content such as playbooks, job templates inside REX, roles and collections on disk, and you relied on modules being delivered by the Ansible RPM on {Project}, you have to take additional steps to adapt your Ansible installation or migrate your Ansible content.

Ansible Core removed all non-essential modules shipped in the core installation.
In terms of FQCN notation `namespace.collection.module`, you can continue to use `ansible.builtin.*`, but everything else is missing in Ansible Core.
That means you will be no longer able to use non-builtin Ansible modules as you were used to and you have to get them from somewhere else, eventually.

Current options to resolve this situation:

> * You can download and install additional community-maintained modules from Galaxy server.

Reviewer 2: ```suggestion
* You can obtain additional community-maintained collections which provide the removed functionality from Ansible Galaxy.
```

* "download and install" → "obtain" is a personal choice of mine (the users don't need to do the download manually, the `ansible-galaxy` tool will do it for them), so change as you see fit
* "modules" → "collections" as that's the more appropriate term for the entity that is installed
* "which provide the removed functionality" - I think that's an important detail ;)

> For more information, see https://galaxy.ansible.com/docs/using/installing.html[Installing content] in _Ansible Galaxy Documentation_.

Reviewer 2: ```suggestion
For more information, see https://docs.ansible.com/ansible/devel/user_guide/collections_using.html#installing-collections[Installing collections] in _Ansible Galaxy Documentation_.
```

The link you have talks about roles, and then links to the above for collections, let's save the readers a click.

ifdef::satellite[]
+
Note that {Team} does not provide support for this content.
endif::[]
ifdef::satellite,orcharhino[]

> * If you have a subscription for {Team} Automation Hub, you can configure your `ansible-galaxy` to talk to Automation Hub server and download content from there.

Reviewer 2: ```suggestion
* If you have a subscription for https://www.ansible.com/products/automation-hub[Ansible Automation Hub], you can configure your `ansible-galaxy` to talk to Automation Hub server and download content from there.
```

Using `{Team}` here is IMHO wrong, as the product name is "Ansible Automation Hub"

That content is supported by {Team}.
For more information on configuring Automation Hub connection for `ansible-galaxy`, see
https://access.redhat.com/documentation/en-us/red_hat_ansible_automation_platform/2.1/html/getting_started_with_automation_hub/proc-configure-automation-hub-server[Configuring Red Hat automation hub as the primary source for content].

> endif::[]

Author: @maximiliankolb Are you sure you want to include it for orcharhino "as is"? E.g. it will render "That content is supported by ATIX"

Reviewer 3: I am unsure right now, so please hide it for orcharhino for now. I might create a follow-up PR in the future.
